### DISCUSSION TITLE
MIDI in not working out of the box

### DISCUSSION THREAD

**Discussion by HansManders** [shorepine/tulipcc#351] [General]
Just received my Tulip. 
After hooking up in my setup, I wanted to play it using an external MIDI keyboard. Connected the Tulip using a TRS cable to an Arturia Minilab 3 but nothing happens. The onscreen keyboard works fine, nice sounds by the way.

Maybe I got the whole idea wrong and I should look at it from a programmer's point of view (which I am, luckily). More like a Norns Shield. But I always thought it would be usable out-of-the-box.

Any ideas?

**Comment by dpwe** [comment]
There are, unfortunately, two standards for a DIN5 to 3.5mm minijack for MIDI; Tulip uses "type A", and the Boss-branded cable I got works for me. 

Before that, I bought a no-name cable claiming to do MIDI, but it was completely wrong (not even type B, it was apparently intended for audio). So your problem may arise from your cable.

**Reply by HansManders** [reply]
Yes, well, that was stupid. The TRS adapter that came with my Iridium Core works like a charm. Thanks!

**Reply by CelloGit** [reply]
Old MIDI keyboard still does not work with the TulipCC (clean boot, nothing running). Some keys work randomly and some notes sound spontaneously but that is it.
I first got a no-name Male Din 5 pin to TRS cable from Amazon that turmed out to be incorrectly wired but changed that to a proper Midi to TRS A cable (Tip (5) = sink, RING (4) = source and SLEEVE (2) = shield)).  Doessss not work though. Any ideas anyone?
Thanks.

**Reply by bwhitman** [reply]
Is there any chance you can buy a "known-good" MIDI-TRS-A connector? That should rule out any issue with the wiring. I have quite a few of the one we recommend and it works fine https://www.amazon.com/Kurrent-Electric-Type-3-5mm-Adapter/dp/B0C2RLB3SL/

**Reply by CelloGit** [reply]
The Kurrent brand cable you are referring to is not available on Amazon.nl. (The Kurrent also has a female DIN connector instead of A male connector. My Yamaha keyaboard has a female MIDI out connector).
I am using this Rockbortd brand type A cable from a reputable music store and checked the wiring: https://www.bax-shop.nl/midi-kabel/rockboard-rbo-cab-f-tma-60-bk-flat-trs-midi-kabel-type-a-60-cm
PS: I am using the ready built TulipCC from Makerfabs and the firmware is up to date
PS2: Compliments mr Brian Whitman on this great project!

**Reply by dpwe** [reply]
I think if we're getting any messages in at all (especially if some of them seem to be correct), the cable is likely to be wired right.  I see two possibilities: Electrical problems (because the keyboard is somehow not sending the correct physical voltages) meaning that the received UART messages are marginal and sometimes garbled, or logical problems because the keyboard is sending MIDI byte sequences that we don't handle well.  

I know we recently substantially hardened the MIDI input byte handling, but I'm wondering if this is still potentially an issue.  Can you tell us the model of keyboard you're using?  And what equipment does it drive successfully?  Ideally, we'd put some kind of protocol analyzer on the keyboard output, but I realize this isn't generally possible!

**Reply by CelloGit** [reply]
The keyboard is a Yamaha PSR-410. MIDI OUT worked fine using my DIN-to-USB converter to connect to a PC but I cannot find that converter at the moment to prove MIDI OUT still works fine.
The other MIDI keyboards I have all use USB connections but finding a (affordable)  USB-to-TRS converter to connect such keyboard to the TulipCC seems to be even a lot more difficult than finding a correctly wired DIN-to-TRS cable.

**Reply by bwhitman** [reply]
Let's try to debug the TRS MIDI. It should work! My first step is to make sure Python is getting the right bytes you're expecting. That's easy on Tulip, just `stop` the default callback and make a new one that prints the hex value of the bytes. Do this in the REPL:

```
>>> midi.stop_default_callback()
>>> def p(m):
...     print(['%02x' % (x) for x in m])
...     
...     
... 
>>> midi.add_callback(p)
```

Then when you hit a key on your keyboard, you should see a note on message come in `0x9X`  (and note off with `0x8X`) with two follow on bytes for note and velocity. What happens when you do this? 

```
['90', '44', '5b']
['80', '44', '40']
```

**Reply by CelloGit** [reply]
1. Managed to verify that MIDI OUT from Yamaha keyboard works fine (found my DIN-to-USB converter).
2. Debugged the DIN-to-TRS MIDI OUT from keyboard to TulipCC by using your code above.
Results are weird. Without touching any key on the keyboard, the input registers (brackest and apostrophes are left out):
c9 00
b1 64 00
b1 65 00
b1 06 02
... some 10 more lines of b1, b0 and c0 sequences
f8
f8
-- this f8 is repeated indefinitely until keyboard is switched off. Actually it then still continues for a while as if the MIDI IN of the TulipCC has been flooded.

This block of code repeats itself when the keyboard is switched on again. Pressing a key on the keyboard rarely succeeds and if it does the MIDI sequence is impossible to spot in the flow of f8's rapidly scrolling on the screen.

thanks

**Reply by bwhitman** [reply]
Great! You're getting valid midi messages. This rules out the hardware (or C-level firmware) 
The F8 repeating is a clock message sent by some controllers. In normal Tulip, we throw it away in `midi.py` before passing it on to the synth. 

For this diagnostic, printing in Tulip is a big CPU operation and the constant printing of the F8 message may be impeding things. So let's try again to see note on messages but not print the F8 clock. 

```python
>>> midi.stop_default_callback()
>>> def p(m):
...     if(len(m)>1):
...         print(['%02x' % (x) for x in m])
...     
...     
... 
>>> midi.add_callback(p)
```

And do it again, get past the Bx/Cx messages, and then hit a key and release it. You should see a single 9x and 8x message if things were working ok.

**Reply by CelloGit** [reply]
Thank you.
Skipping the f8 reveals better data.
What we now register when pressing a key is 90 xx yy!
But releasing the key does not send any code (80 ww zz expected).
Moreover: 
1. it takes a while before a next keypress is registered. CPU stress?
2. in 90 xx yy, xx is velocity and yy is note instead of the other way around as you write above.

**Reply by bwhitman** [reply]
Interesting! I made a small change to Tulip to stop sending `F8` to Python for processing. Tulip should either handle F8 at the firmware (to set tempo/clock) or ignore it. My _hypothesis_ is that fast F8s from certain controllers are flooding the Python call stack. 

Can you do `tulip.upgrade()`, and make sure when it reboots the screen says `2024-09-24`, and then try again to play midi notes?

**Reply by CelloGit** [reply]
Good morning!
Alas. Upgraded the TulipCC but did not solve issue.

**Reply by CelloGit** [reply]
Double checked by connecting a PC to the TulipCC via my DIN-to-USB/USB-to-DIN converter, hooking up an USB MIDI keyboard to the PC and connecting the keyboard to the converter via the nice MIDIberry software.
Ran your "def p(m)" routine and got the correct 90 xx yy and 80 vv ww sequences displayed without any delays.
thanks

**Reply by bwhitman** [reply]
Ok! So you're saying sending midi messages to tulip works via your computer but not directly wired to your controller? 

can you try running that diagnostic again on your upgraded tulip so we can see all the messages ? You should no longer see f8s. But I'm curious to see what else is coming in when you play notes.

**Reply by bwhitman** [reply]
(I did a little googling and found this reddit thread, so maybe there's something "non-standard" about the PSR-410 that we have to learn about! https://www.reddit.com/r/midi/comments/q02djc/yamaha_psr410_midi_issue/ )

**Reply by CelloGit** [reply]
Results from three setups on the firmware updated TulipCC are in photo below.

>>> yamaha via pc

yahama PSR-410 OUT is connected to PC via two-way smart DIN/USB adapter (called AdX hereafter).

PC is connected to TulipCC IN via AdX + simple DIN-to-TSR connector.

In this setup everything sound fine and smooth on PC soft synth and TulipCC juno (nice!).
Readout is strange though since it has no 80 vv ww sequences.

>>> minilab via pc

Arturia Minilab MIDI keyboard connected to pc via usb.

PC is connected to TulipCC IN via AdX + stupid DIN-to-TSR connector.

In this setup everything sounds fine and smooth on PC soft synth and on TulipCC juno (nice!).
Readout shows both 90 xx yy and 80 vv ww sequences.

>>> yamaha direct

Yamaha OUT is drectly connected to TulipCC IN via simple DIN-to-TRS Type A cable.
We get same behaviour as on TulipCC before the 24/9/24 firmware update.
Readout shows only 90 xx yy sequences, but much worse: as before CPU is stressed since keypresses are only recognized randomly and much delayed. And lots of keypresses are not recognized at all.

![20240925_151430_cr](https://github.com/user-attachments/assets/a21a5632-ee56-4abf-bdb0-eedb48c7bbe2)

**Reply by CelloGit** [reply]
Eureka, it dawned on me!

The USB keyboard input of the TulipCC also accepts USB MIDI keyboards. Kudos to the designers.  
So the issue described above of the Yamaha DIN MIDI keyboard not behaving nicely with the TRS input has become much less urgent.

A drawback of using the USB input is of course that it allows only one USB device connected at the same time. Maybe in the future a hub will be supported. In the meantime I should perhaps build a mechanical switch as alternative to continously swapping cables on the USB input. Or getting up to speed typing on the M5Stack CardKB. But that is a real challenge.

PS. Apologies for the somewhat indecent large size of the screenshot above.

**Comment by znmeb** [comment]
The ones I get from Amazon are color-coded - bright red for type B and bright blue for type A:

https://www.amazon.com/gp/product/B08TM8BT3G/?th=1
https://www.amazon.com/gp/product/B08WHSP7ZL/?th=1
